File: src/main/kotlin/com/example/demo/ClinicalNoteGenerator.java
package com.example.demo;

import okhttp3.*;
import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;

import java.io.IOException;

public class ClinicalNoteGenerator {

    private static final String API_KEY = "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";
    private static final String API_URL = "https://api.openai.com/v1/chat/completions";

    public static void main(String[] args) {
        String transcription = "Patient reports severe headaches for the last two weeks, currently taking ibuprofen daily.";

        try {
            String summary = generateClinicalNoteFromTranscription(transcription);
            System.out.println("Generated Clinical Note:\n" + summary);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static String generateClinicalNoteFromTranscription(String transcription) throws IOException {
        OkHttpClient client = new OkHttpClient();

        // Create JSON body
        String json = new Gson().toJson(new ChatRequest(transcription));

        // Build request
        RequestBody body = RequestBody.create(json, MediaType.get("application/json; charset=utf-8"));
        Request request = new Request.Builder()
                .url(API_URL)
                .post(body)
                .addHeader("Authorization", "Bearer " + API_KEY)
                .build();

        // Execute request
        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Unexpected code " + response);
            }

            String responseBody = response.body().string();
            ChatResponse chatResponse = new Gson().fromJson(responseBody, ChatResponse.class);

            // Assuming response contains a list of choices
            if (chatResponse.choices != null && chatResponse.choices.length > 0) {
                return chatResponse.choices[0].message.content;
            } else {
                return "No content found in the response.";
            }
        }
    }

    // Inner classes to represent the request and response structure
    static class ChatRequest {
        String model = "gpt-4";
        ChatMessage[] messages;

        ChatRequest(String transcription) {
            this.messages = new ChatMessage[]{new ChatMessage("user", transcription)};
        }
    }

    static class ChatMessage {
        String role;
        String content;

        ChatMessage(String role, String content) {
            this.role = role;
            this.content = content;
        }
    }

    static class ChatResponse {
        Choice[] choices;

        static class Choice {
            Message message;

            static class Message {
                @SerializedName("content")
                String content;
            }
        }
    }
}

File: src/main/kotlin/com/example/demo/ClinicalNoteGenerator.kt
package com.example.demo
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File
import java.io.IOException

fun ClinicalMain() {
    val transcriptionFilePath = "transcribed files/7087_transcription_result.txt"
    val gptApiKey = "passvalidtokenhere"
    val gptApiUrl = "https://api.openai.com/v1/chat/completions"

    val transcriptionText = try {
        File(transcriptionFilePath).readText(Charsets.UTF_8)
    } catch (e: IOException) {
        println("Error reading transcription file: ${e.message}")
        return
    }

    // Generate clinical notes from transcription
    val client = OkHttpClient()

    val requestBody = Gson().toJson(
        mapOf(
            "model" to "gpt-4",
            "messages" to listOf(
                mapOf(
                    "role" to "user",
                    "content" to "Generate clinical notes based on the following transcription:\n\n$transcriptionText"
                )
            ),
            "max_tokens" to 150,
            "temperature" to 0.7
        )
    )

    val request = Request.Builder()
        .url(gptApiUrl)
        .post(requestBody.toRequestBody("application/json".toMediaType()))
        .addHeader("Authorization", "Bearer $gptApiKey")
        .addHeader("Content-Type", "application/json")
        .build()

    try {
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Unexpected code $response")
            }

            val responseBody = response.body?.string() ?: ""
            val gson = Gson()
            val mapType = object : TypeToken<Map<String, Any>>() {}.type
            val gptResponse = gson.fromJson<Map<String, Any>>(responseBody, mapType)

            // Safely parse the response
            val choices = gptResponse["choices"] as? List<*>
            val choice = choices?.firstOrNull() as? Map<*, *>
            val content = (choice?.get("message") as? Map<*, *>)?.get("content") as? String
            val clinicalNotes = content ?: "No clinical notes generated."

            println("Generated Clinical Notes:\n$clinicalNotes")
        }
    } catch (e: IOException) {
        println("Error generating clinical notes: ${e.message}")
    }
}